README.md
# Monitor Laravel Queue and Schedule

This package runs a queued job and a scheduled command once every 5 minutes and a route to check that this has happened successfully.

This route will return a 200 or a 503 HTTP status which is useful for monitoring software such as StatusCake.

## Limitations

This package relies on the database queue driver being used.

## Installation

Install the package via composer:

```bash
composer require hellomayaagency/laravel-heartbeat
```

Enable the scheduled heartbeat command:

```php
// app/Console/Kernel.php

protected function schedule(Schedule $schedule)
{
    // ...

    // Note: don't change the interval!
    $schedule->command('heartbeat')->everyFiveMinutes();
}
```

Run the heartbeat command:

```bash
php artisan heartbeat
```

Add a route to the heartbeat status page:

```php
// routes/web.php
Route::get('heartbeat', '\Maya\LaravelHeartbeat\Http\Controllers\HeartbeatController@index');

// or

Route::get('heartbeat', [HeartbeatController::class, 'index');
```

## Usage

Navigate to the route you created above, e.g. `example.com/heartbeat`. You should see a status page.

The status page will return a 200 status page if everything is working as expected.

If problems are found with the schedule or queue a 503 status will be returned instead.

### Testing

```bash
composer test
```

### Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

### Security

If you discover any security related issues, please email jake@maya.agency instead of using the issue tracker.

## Credits

-   [Jake Gully](https://github.com/hellomayaagency)
-   [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

## Laravel Package Boilerplate

This package was generated using the [Laravel Package Boilerplate](https://laravelpackageboilerplate.com).


CHANGELOG.md
# Changelog

All notable changes to `laravel-heartbeat` will be documented in this file

## 1.0.0 - 201X-XX-XX

- initial release


config/config.php
<?php

/*
 * You can place your custom package configuration in here.
 */
return [
    /*
    |--------------------------------------------------------------------------
    | Heartbeat Interval Time
    |--------------------------------------------------------------------------
    |
    | Time in minutes between heartbeat jobs. This will confirm that
    | the queue is working correctly and will make the status page
    | return 200
    */
    'heartrate' => 10,

    /*
    |--------------------------------------------------------------------------
    | Heartbeat status page layout
    |--------------------------------------------------------------------------
    |
    | The heartbeat status page will extend this view. Defaults to
    | 'layouts.app`.
    */
    'layout' => 'layouts.app',
];


src/Console/Heartbeat.php
<?php

namespace Maya\LaravelHeartbeat\Console;

use Carbon\Carbon;
use Illuminate\Console\Command;
use Illuminate\Support\Facades\Cache;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Log;
use Maya\LaravelHeartbeat\Jobs\QueueHeartbeat;

class Heartbeat extends Command
{
    /**
     * The name and signature of the console command.
     *
     * @var string
     */
    protected $signature = 'heartbeat';

    /**
     * The console command description.
     *
     * @var string
     */
    protected $description = 'Signal that application is running';

    /**
     * Create a new command instance.
     *
     * @return void
     */
    public function __construct()
    {
        parent::__construct();
    }

    /**
     * Execute the console command.
     *
     * @return mixed
     */
    public function handle()
    {
        Log::info('heartbeat');

        Cache::forever('schedule_heartbeat', Carbon::now());

        $job_count = DB::table('jobs')->where('payload', 'like', '%QueueHeartbeat%')->count('id');

        if ($job_count === 0) {
            Log::info('heartbeat - Queued QueueHeartbeat');

            QueueHeartbeat::dispatch()->delay(Carbon::now()->addMinute());
        }
    }
}


src/Jobs/QueueHeartbeat.php
<?php

namespace Maya\LaravelHeartbeat\Jobs;

use Carbon\Carbon;
use Illuminate\Bus\Queueable;
use Illuminate\Contracts\Queue\ShouldQueue;
use Illuminate\Foundation\Bus\Dispatchable;
use Illuminate\Queue\InteractsWithQueue;
use Illuminate\Queue\SerializesModels;
use Illuminate\Support\Facades\Cache;
use Illuminate\Support\Facades\Log;

class QueueHeartbeat implements ShouldQueue
{
    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;

    /**
     * Create a new job instance.
     *
     * @return void
     */
    public function __construct()
    {
        //
    }

    /**
     * Execute the job.
     *
     * @return void
     */
    public function handle()
    {
        Log::info('queue heartbeat');

        Cache::forever('queue_heartbeat', Carbon::now());

        static::dispatch()->delay(Carbon::now()->addMinutes(config('app.heartrate')));
    }
}


resources/views/status.blade.php
@extends(config('heartbeat.layout','layouts.app'))

@section('content')
<div class="container">

  <section class="section">
    <h1 class="title is-1">Status</h1>
    @if ($schedule_ok)
      <h2 class="title is-3">
        <span class="has-text-success">✓</span> Schedule is ok!
      </h2>
    @else
      <h2 class="title is-3">
        <span class="has-text-danger">✘</span> Schedule is not ok!
      </h2>
    @endif

    <p>
      @if (is_null($last_schedule))
        <span class="has-text-danger">✘</span> No record of last schedule heartbeat.
      @else
        @if ($schedule_ok)
          <span class="has-text-success">✓</span>
        @else
          <span class="has-text-danger">✘</span>
        @endif
        Last schedule heartbeat was {{ $last_schedule->toDateTimeString() }} - {{ $last_schedule->diffForHumans() }}
      @endif
    </p>
  </section>

  <section class="section">
    @if ($queue_ok)
      <h2 class="title is-3">
        <span class="has-text-success">✓</span> Queue is ok!
      </h2>
    @else
      <h2 class="title is-3">
        <span class="has-text-danger">✘</span> Queue is not ok!
      </h2>
    @endif

    <p>
      @if (is_null($last_queue))
        <span class="has-text-danger">✘</span> No record of last queue heartbeat.
      @else
        @if($queue_ok)
          <span class="has-text-success">✓</span>
        @else
          <span class="has-text-danger">✘</span>
        @endif
        Last queue heartbeat was {{ $last_queue->toDateTimeString() }} - {{ $last_queue->diffForHumans() }}
      @endif
    </p>

    @if ($job)
      <p>
        <span class="has-text-success">✓</span> Next job is scheduled for {{ \Carbon\Carbon::createFromTimestampUTC($job->available_at)->toDateTimeString() }} - {{ \Carbon\Carbon::createFromTimestampUTC($job->available_at)->diffForHumans() }}
      </p>
    @else
      <p>
        <span class="has-text-danger">✘</span> There is no heartbeat job in the queue. Try: <code>php artisan heartbeat</code>
      </p>
    @endif
  </section>
</div>

@endsection


src/LaravelHeartbeatServiceProvider.php
<?php

namespace Maya\LaravelHeartbeat;

use Illuminate\Support\ServiceProvider;
use Maya\LaravelHeartbeat\Console\Heartbeat;

class LaravelHeartbeatServiceProvider extends ServiceProvider
{
    /**
     * Bootstrap the application services.
     */
    public function boot()
    {
        /*
         * Optional methods to load your package assets
         */
        // $this->loadTranslationsFrom(__DIR__.'/../resources/lang', 'laravel-heartbeat');
        // $this->loadViewsFrom(__DIR__.'/../resources/views', 'laravel-heartbeat');
        // $this->loadMigrationsFrom(__DIR__.'/../database/migrations');
        // $this->loadRoutesFrom(__DIR__.'/routes.php');

        $this->loadViewsFrom(__DIR__.'/../resources/views', 'laravel-heartbeat');

        if ($this->app->runningInConsole()) {
            $this->publishes([
                __DIR__.'/../config/config.php' => config_path('laravel-heartbeat.php'),
            ], 'config');

            // Publishing the views.
            // $this->publishes([
            //     __DIR__.'/../resources/views' => resource_path('views/vendor/laravel-heartbeat'),
            // ], 'views');

            // Publishing assets.
            /*$this->publishes([
                __DIR__.'/../resources/assets' => public_path('vendor/laravel-heartbeat'),
            ], 'assets');*/

            // Publishing the translation files.
            /*$this->publishes([
                __DIR__.'/../resources/lang' => resource_path('lang/vendor/laravel-heartbeat'),
            ], 'lang');*/

            // Registering package commands.
            $this->commands([
                Heartbeat::class,
            ]);
        }
    }

    /**
     * Register the application services.
     */
    public function register()
    {
        // Automatically apply the package configuration
        $this->mergeConfigFrom(__DIR__.'/../config/config.php', 'laravel-heartbeat');

        // Register the main class to use with the facade
        $this->app->singleton('laravel-heartbeat', function () {
            return new LaravelHeartbeat;
        });
    }
}


src/Http/Controllers/HeartbeatController.php
<?php

namespace Maya\LaravelHeartbeat\Http\Controllers;

use Carbon\Carbon;
use Illuminate\Support\Facades\Cache;
use Illuminate\Support\Facades\DB;

class HeartbeatController
{
    /**
     * Get the health status of the application
     *
     * @return View
     */
    public function index()
    {
        $now = Carbon::now();
        $last_schedule = Cache::get('schedule_heartbeat');
        $time_since_last_schedule = null;
        $schedule_ok = false;

        // We'll allow an extra bit of time between heartbeats and checks
        // to prevent any race conditions.
        $leeway = 1;

        if ($last_schedule) {
            $schedule_interval_mins = config('heartbeat.heartrate') + $leeway;
            $time_since_last_schedule = $now->diffInMinutes($last_schedule);
            $schedule_ok = $time_since_last_schedule < $schedule_interval_mins;
        }

        $last_queue = Cache::get('queue_heartbeat');
        $time_since_last_queue = null;
        $queue_ok = false;

        if ($last_queue) {
            $queue_interval_mins = config('heartbeat.heartrate') + $leeway;
            $time_since_last_queue = $now->diffInMinutes($last_queue);
            $queue_ok = $time_since_last_queue < $queue_interval_mins;
        }

        $job = DB::table('jobs')->where('payload', 'like', '%QueueHeartbeat%')->orderBy('available_at', 'ASC')->first();

        $status = $schedule_ok && $queue_ok ? 200 : 503;

        return response()
            ->view('laravel-heartbeat::status', compact(
                'schedule_ok',
                'last_schedule',
                'time_since_last_schedule',
                'queue_ok',
                'last_queue',
                'time_since_last_queue',
                'job'
            ), $status);
    }
}
